vault/namespaces_store.go

```diff
@@ -5,70 +5,167 @@ package vault
 
 import (
 	"context"
+	"errors"
 	"fmt"
-	"strings"
+	"path"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/armon/go-metrics"
 	"github.com/hashicorp/go-hclog"
-	"github.com/hashicorp/go-secure-stdlib/strutil"
+	"github.com/hashicorp/go-secure-stdlib/base62"
+	uuid "github.com/hashicorp/go-uuid"
 	"github.com/openbao/openbao/helper/namespace"
 	"github.com/openbao/openbao/sdk/v2/logical"
 )
 
-func (c *Core) NamespaceByID(ctx context.Context, nsID string) (*namespace.Namespace, error) {
-	return namespaceByID(ctx, nsID, c)
-}
-
-func (c *Core) ListNamespaces(includePath bool) []*namespace.Namespace {
-	return []*namespace.Namespace{namespace.RootNamespace}
-}
-
-func (c *Core) resetNamespaceCache() {}
+// Namespace id length; upstream uses 5 characters so we use one more to
+// differentiate OpenBao from Vault Enterprise. This allows 2^35 entries; with
+// fairly high probability, we'll hit a conflict here and have to regenerate
+// but we shouldn't ever run out. This is also different from mount accessors
+// (8 hex characters).
+//
+// See: https://developer.hashicorp.com/vault/api-docs/system/namespaces
+const namespaceIdLength = 6
 
-// start all new namespace code below
+// Namespace storage location.
+const namespaceStoreRoot = "core/namespaces/"
 
-var immutableNamespaces = []string{
-	"sys",
-	"audit",
-	"auth",
-	"cubbyhole",
-	"identity",
-}
-
-// NamespaceStore is used to provide durable storage of namespace
+// NamespaceStore is used to provide durable storage of namespace. It is
+// a singleton store across the Core and contains all child namespaces.
 type NamespaceStore struct {
 	core    *Core
-	aclView BarrierView
+	storage logical.Storage
+
+	// This lock ensures we don't concurrently modify the store while using
+	// a namespace entry. We also store an atomic to check if we need to
+	// reload all namespaces.
+	lock        sync.RWMutex
+	invalidated atomic.Bool
 
-	// This is used to ensure that writes to the store (acl) or to the egp
-	// path tree don't happen concurrently. We are okay reading stale data so
-	// long as there aren't concurrent writes.
-	modifyLock *sync.RWMutex
+	// List of all namespaces within the store. This is loaded at store
+	// initialization time and persisted throughout the lifetime of the
+	// instance. Entries should not be returned directly but instead be
+	// copied to prevent modification.
+	namespaces []*NamespaceEntry
 
 	// logger is the server logger copied over from core
 	logger hclog.Logger
 }
 
-// NamespaceEntry is used to store a namespace by name
+// NamespaceEntry is used to store a namespace. We wrap namespace.Namespace
+// in case there is additional data we wish to store that isn't relevant to
+// a namespace instance.
 type NamespaceEntry struct {
-	namespace.Namespace
+	UUID      string               `json:"uuid"`
+	Namespace *namespace.Namespace `json:"namespace"`
+}
```

Comment:
According to the RFC we also want to have a field `parent`:

> `parent`, the uuid of the parent namespace of this namespace and not returned from the response API.

I guess we would add this one here?

```suggestion
	UUID      string               `json:"uuid"`
	Namespace *namespace.Namespace `json:"namespace"`
	Parent      string
}
```

Reply:
Yeah, I think @driif had filed #1022 which has: 

> 2. Parent Namespace Validation
>     Currently, it's possible to create namespaces under non-existing parent namespaces:

and in general the first iteration didn't really support hierarchical namespaces properly. I think we should probably split that point out as a separate ticket and add `parent` as an explicit separate PR.

Reply:
Ahh, I think I remember (keep me honest @genelet / @klaus-sap) -- we had originally thought of a flat namespace store, strictly under `/core/namespaces/<uuid>`. Child namespaces wouldn't be under `/namespace/<parent-uuid>/core/namespaces/<child-uuid>` but instead all part of the global root store.

Curious what we feel about either design approach? Parent would be necessary in the former case, but the latter would keep all of a parent namespace's data in one place.

Reply:
@cipherboy `/namespace/<parent-uuid>/core/namespaces/<child-uuid>` feels weird to me. If we'd go hierarchical, I think the way it would make the most sense is just using the (already hierarchical) path instead of uuids (`/core/namespaces/<parent-name>/<child-name>`, so the equivalent to `/core/namespaces/<full-namespace-path>`).

Otherwise I'd go with the flat hierarchy with uuids.

Reply:
I would prefer a flat hierarchy with uuids.

Reply:
@phyrog @klaus-sap Note that in the additional work section, we have:
 
https://github.com/openbao/openbao/issues/787

> Per-Namespace Storage Segments

If we place all of this information in (root) Core, we mix between a self-contained namespace and not. Where do you draw the line and why stop at the namespace table? Wouldn't it (by similar logic) make sense to keep the former mount table implementation, which is a single global entry, and not move mount information into each namespace? And leases... And tokens... And ACLs... &c.

With the interactions with lazy-loading and seal/unseal of namespaces (which are closely related), you end up with a root namespaces which has references to namespaces which don't have storage mounted/accessible/may be broken/... and so have to continually skip them, despite them contributing to LIST operations for the above operations (namespaces, token store tidying, lease tidying, &c). 

For scalability, my opinion is that it is better to push, consistently, all of this information into the relevant parent namespace and not clutter root.

Though, I'd probably suggest a single-level hierarchy and say all namespaces, regardless of type, should be rooted at `/namespace/<uuid>` so the path doesn't grow too long (which impacts performance on Psql or File backend). So `/namespaces/<parent-uuid>` and `/namespaces/<child-uuid>` are siblings in the hierarchy. 

That said, the use of UUID aligns with the mount table and `logical` barrier view. I prefer this as we hope to reduce information leakage and paths aren't necessarily guaranteed to be encrypted. So I'd avoid using `path` there, even if we're sure it wouldn't cause injection issues. 

My 2c.